crowbar_framework/app/models/neutron_service.rb

```diff
@@ -94,14 +94,40 @@ def validate_proposal_after_save proposal
     end
 
     plugin = proposal["attributes"]["neutron"]["networking_plugin"]
-    mode = proposal["attributes"]["neutron"]["networking_mode"]
+    ml2_mechanism_drivers = proposal["attributes"]["neutron"]["ml2_mechanism_drivers"]
+    ml2_type_drivers = proposal["attributes"]["neutron"]["ml2_type_drivers"]
+    ml2_type_drivers_default_provider_network = proposal["attributes"]["neutron"]["ml2_type_drivers_default_provider_network"]
+    ml2_type_drivers_default_tenant_network = proposal["attributes"]["neutron"]["ml2_type_drivers_default_tenant_network"]
+
+    # default provider network ml2 type driver must be a driver from the selected ml2 type drivers
+    # TODO(toabctl): select the ml2_type_driver automatically if used as default? Or directly check in the ui via js?
+    unless ml2_type_drivers.include?(ml2_type_drivers_default_provider_network)
+      validation_error("The default provider network type driver \"#{ml2_type_drivers_default_provider_network}\" is not a selected ml2 type driver")
+    end
+
+    # default tenant network ml2 type driver must be a driver from the selected ml2 type drivers
+    # TODO(toabctl): select the ml2_type_driver automatically if used as default? Or directly check in the ui via js?
+    unless ml2_type_drivers.include?(ml2_type_drivers_default_tenant_network)
+      validation_error("The default tenant network type driver \"#{ml2_type_drivers_default_tenant_network}\" is not a selected ml2 type driver")
+    end
 
-    if %(linuxbridge cisco).include?(plugin) and mode != "vlan"
-      validation_error("The \"#{plugin}\" plugin only supports the mode: \"vlan\"")
+    # linuxbridge and cisco_nexus mech drivers need vlan type driver
+    # TODO(toabctl): select vlan type driver automatically if linuxbridge or cisco were selected!?
+    %w(linuxbridge cisco_nexus).each do |drv|
+      if ml2_mechanism_drivers.include? drv and not ml2_type_drivers.include? 'vlan'
+        validation_error("The mechanism driver \"#{drv}\" needs the type driver \"vlan\"")
+      end
     end
 
-    if !%(gre vlan).include?(mode)
-      validation_error("Unknown networking mode \"#{mode}\"")
+    # cisco_nexus mech driver needs also openvswitch mech driver
+    # TODO(toabctl): select openvswitch automatically if cisco_nexus was selected!?
+    if ml2_mechanism_drivers.include? "cisco_nexus" and not ml2_mechanism_drivers.include? "openvswitch"
+      validation_error("The 'cisco_nexus' mechanism driver needs also the 'openvswitch' mechanism driver")
+    end
+
+    # for now, openvswitch and linuxbrige can't be used in parallel
+    if ml2_mechanism_drivers.include? "openvswitch" and ml2_mechanism_drivers.include? "linuxbridge"
+      validation_error("The 'openvswitch' and 'linuxbridge' mechanism drivers can't be used in parallel. Only select one of them")
     end
```

Comment:
> Shouldn't we validate that ml2_mechanism_drivers should have at least one of openvswitch or linuxbridge?
> 
> Similarly, for ml2_type_drivers, shouldn't we check it has at least vlan or gre?

That's already checked because we set 'required' to True in the barclamp schema.